Useful helper functions for Unstructured

ash2k opened the pull request:
**Which issue this PR fixes**:
Fixes #40790

**Release note**:
```release-note
NONE
```
/kind feature
/sig api-machinery
/area client-libraries
/assign @sttts @liggitt

sttts commented:
/ok-to-test

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -200,22 +201,28 @@ func getNestedField(obj map[string]interface{}, fields ...string) interface{} {
-	nested := getNestedField(obj, fields...)
+// NestedInt64Pointer returns the *int64 value of a nested field.
+// Returns nil if value is not found or is not an int64/*int64.
+func NestedInt64Pointer(obj map[string]interface{}, fields ...string) *int64 {
```
why a pointer and not the value?

sttts replied:
The other funcs return values or clones, never the original object or a reference to it.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -200,22 +201,28 @@ func getNestedField(obj map[string]interface{}, fields ...string) interface{} {
+// NestedString returns the string value of a nested field.
+// Returns empty string if value is not found or is not a string.
+func NestedString(obj map[string]interface{}, fields ...string) string {
+	if str, ok := NestedField(obj, fields...).(string); ok {
```
I would prefer to see `error`s in the result if the type is wrong.

Same for the other type-casts below.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -226,8 +233,10 @@ func getNestedInt64Pointer(obj map[string]interface{}, fields ...string) *int64
+	if m, ok := NestedField(obj, fields...).([]interface{}); ok {
 		strSlice := make([]string, 0, len(m))
 		for _, v := range m {
 			if str, ok := v.(string); ok {
```
I would prefer to return an error if the cast fails.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -239,8 +248,21 @@ func getNestedSlice(obj map[string]interface{}, fields ...string) []string {
+	if m, ok := NestedField(obj, fields...).(map[string]interface{}); ok {
 		strMap := make(map[string]string, len(m))
 		for k, v := range m {
 			if str, ok := v.(string); ok {
```
same here, I prefer an error in this case

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -252,106 +274,149 @@ func getNestedMap(obj map[string]interface{}, fields ...string) map[string]strin
-			m = m[field].(map[string]interface{})
+	for _, field := range fields[:len(fields)-1] {
+		if _, ok := m[field].(map[string]interface{}); !ok {
+			m[field] = make(map[string]interface{})
```
this is unexpected that everything of wrong type is overridden.

sttts replied:
Also here we need errors.

ash2k replied:
Well, I agree with you. This is how it is implemented right now, I just made it public. I'm happy to change it.
Do you want all of these functions to return an `error`? Maybe `bool` for getters?

sttts replied:
`bool` is also fine, maybe even better.

lavalamp commented:
/assign @krousey

ash2k committed:
Split unstructured.go into several parts

No functional changes

ash2k commented:
@sttts I've updated the PR. I've split `Unstructured.go` into several files - it was too big of a file at ~1k lines. I hope it's ok. Second commit makes the necessary changes.

I've changed some methods to return a `bool` if value was not found or cannot be retrieved (or set) because existing data is of invalid type. If we want to distinguish between those cases then we have 3 situations - valid response, value not found, data is invalid (of wrong type). What should we do, WDYT?
I don't think adding third return value is worth it (e.g. value interface{}, found bool, errIfInvalid error). If someone needs precise control of what is happening they can always roll their own functions. However we can add logging to places where we detect invalid data types.
I have also added assertion to catch attempts to set value of an invalid type.

ash2k commented:
Another thing - I don't think making shallow copies on gets/sets makes sense. Either deep copies should be done or no copying at all. I've left the current behaviour for now to get your opinion first.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -0,0 +1,378 @@
+	return in, nil
+}
+
+func (UnstructuredObjectConverter) ConvertFieldLabel(version, kind, label, value string) (string, string, error) {
```
are these change pure copies?

sttts replied:
if not, two separate commits would help.

ash2k replied:
There are two commits, first splits file into multiple without modifications, second one makes changes. Yes, these "changes" are not changes, just copying.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -31,171 +31,278 @@ import (
 
-func getNestedField(obj map[string]interface{}, fields ...string) interface{} {
+// NestedField returns the value of a nested field.
+// false is returned if the value is missing.
```
So `null` gives `nil, true`, undefined gives `nil, false` ? Maybe be a bit more verbose here.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -31,171 +31,278 @@ import (
 			if str, ok := v.(string); ok {
 				strSlice = append(strSlice, str)
+			} else {
+				return nil, false
```
This is surprising. So the first non-string value in the slice will lead to `nil, false`?

sttts replied:
at least the godoc does not match

sttts replied:
this is still open

sttts replied:
fixed in 3rd commit :+1:

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -31,171 +31,278 @@ import (
 			if str, ok := v.(string); ok {
 				strMap[k] = str
+			} else {
+				return nil, false
```
godoc does not match

sttts replied:
still open

sttts replied:
fixed in 3rd commit :+1:

sttts commented:
> Another thing - I don't think making shallow copies on gets/sets makes sense. Either deep copies should be done or no copying at all. I've left the current behaviour for now to get your opinion first.

Have you checked where the helpers are called? I agree that deepcopies would probably be right.

liggitt commented:
> I don't think making shallow copies on gets/sets makes sense. Either deep copies should be done or no copying at all

are you referring to this:
```
		listObj := make(map[string]interface{}, len(t.Object)+1)
		for k, v := range t.Object { // Make a shallow copy
			listObj[k] = v
		}
		listObj["items"] = items
		return json.NewEncoder(w).Encode(listObj)
```

A shallow copy is required, a deep copy is not (and is significantly more expensive)

sttts commented:
> are you referring to this:

I though he referred to https://github.com/kubernetes/kubernetes/pull/51940/files#diff-5ea811de4cbc97262ff557ce638b82b6R125 and the one in SetNestedSlice.

ash2k commented:
@sttts Addressed comments, please take a look at the last commit. Added more detailed godocs, added proper deep copying, added logging in few places where data type is unexpected.

I noticed a possible bug - take a look at `SetInitializers()` and `GetInitializers()`. After to/from unstructured conversion fails they don't return from the method, continuing instead. So `GetInitializers()` returns a non-nil pointer with potentially invalid data and `SetInitializers()` sets the field to `nil`/invalid value. Is that desired?

ash2k commented:
@sttts PTAL

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -0,0 +1,472 @@
+// NestedField returns a deep copy of the value of a nested field.
+// false is returned if the value is missing.
+// nil, true is returned for a nil field.
+func NestedField(obj map[string]interface{}, fields ...string) (interface{}, bool) {
```
`NestedFieldCopy` would be less surprising.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -42,6 +53,7 @@ func NestedField(obj map[string]interface{}, fields ...string) (interface{}, boo
 				return nil, false
 			}
 		} else {
+			glog.Warningf("NestedField(): expected map[string]interface{}, got %T while getting field %q in %s", val, field, fields)
```
don't think we want this warning. I would expect that there are ways to trigger this warning from outside of the apiserver, spamming the log.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -426,7 +413,7 @@ func (s unstructuredJSONScheme) decodeToList(data []byte, list *UnstructuredList
 
 	delete(list.Object, "items")
-	list.Items = nil
+	list.Items = make([]Unstructured, 0, len(dList.Items))
```
empty `dList.Items` gave nil before, but [] now. Not sure this matters. But I would probably keep the old behaviour to avoid regressions which are hard to find.

ash2k replied:
`UnstructuredList.DeepCopy()` does this today already so should not be a problem:
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L158-L170

I've also checked all places where this field is used - no issues. JSON encoding should not change too - it always produces an empty array:
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L692-L706

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -150,6 +152,7 @@ func (u *Unstructured) GetOwnerReferences() []metav1.OwnerReference {
 	for _, obj := range original {
 		o, ok := obj.(map[string]interface{})
 		if !ok {
+			glog.Warningf("GetOwnerReferences(): expected map[string]interface{}, got %T", obj)
```
As before, not sure about these warnings.

sttts commented:
> I noticed a possible bug - take a look at SetInitializers() and GetInitializers(). After to/from unstructured conversion fails they don't return from the method, continuing instead. So GetInitializers() returns a non-nil pointer with potentially invalid data and SetInitializers() sets the field to nil/invalid value. Is that desired?

This untyped code is full of corner cases if unexpected/invalid data is injected, choosing some (hopefully) neutral default in those cases. I don't think `Set/GetInitializer` is special here. Unfortunately, Go cannot help us much.

sttts commented:
@ash2k the only non-nit issue are the warnings. If we want to keep them, I guess `runtime.HandleError` would be the right way to output them. I tend to remove them though. @liggitt opinions?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -29,11 +29,22 @@ import (
+	if !ok {
+		return nil, false
+	}
+	return unstructured.DeepCopyJSONValue(val), true
```
why are we deep copying here? need to double check this doesn't impact the uses by patch. cc @wojtek-t

wojtek-t replied:
I think it probably will impact.

Before proceeding with this change (in general with this PR), I would like to see the impact of those changes to this benchmark:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unstructured_test.go#L134

ash2k replied:
Deep copying only matters (and is done in this PR) for slices and maps. Both `getNestedSlice()` and `getNestedMap()` in `master` currently make a copy: 
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L209-L233
Another place where shallow copying is done currently is `getOwnerReferences()` - a helper function used by corresponding methods on Unstructured(List):
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L333-L356
This is actually unnecessary because `GetOwnerReferences()` method converts data into structs using `extractOwnerReference ()`:
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L358-L369
https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L285-L313

I have removed this copying.
Regarding the benchmark - looks like there is no difference (probably because there is not much extra copying happening):
```
on master
/usr/local/Cellar/go/1.9.1/libexec/bin/go test -c -i -o /private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go k8s.io/kubernetes/pkg/api/testing
/private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go -test.v -test.bench ^BenchmarkToFromUnstructured$ -test.run ^$
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/api/testing
20000	     79458 ns/op
PASS

Process finished with exit code 0
```
```
this PR:
/usr/local/Cellar/go/1.9.1/libexec/bin/go test -c -i -o /private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go k8s.io/kubernetes/pkg/api/testing
/private/var/folders/wk/_45zctx51y57kvc6sytpfmm40000gn/T/___BenchmarkToFromUnstructured_in_unstructured_test_go -test.v -test.bench ^BenchmarkToFromUnstructured$ -test.run ^$
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/api/testing
20000	     79527 ns/op
PASS

Process finished with exit code 0
```
Actually, I have reviewed usages of this method and replaced two places where it was used with unexported `nestedFieldNoCopy()`. So this change should have 0 impact.

ash2k commented:
@sttts @liggitt @wojtek-t I have addressed all comments - please take a look at the last commit. Removed logging and removed unnecessary copying. Also answered some review comments.

ash2k commented:
/retest

ash2k commented:
@sttts Would be great to get this merged into 1.9, PTAL.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -262,31 +271,47 @@ func (u *Unstructured) GetDeletionTimestamp() *metav1.Time {
 
 func (u *Unstructured) GetDeletionGracePeriodSeconds() *int64 {
-	return getNestedInt64Pointer(u.Object, "metadata", "deletionGracePeriodSeconds")
+	val, ok := NestedInt64(u.Object, "metadata", "deletionGracePeriodSeconds")
```
related https://github.com/kubernetes/kubernetes/pull/54803

sttts commented:
Can you squash commit 2, 3 and 5?

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -31,171 +31,278 @@ import (
-func setNestedField(obj map[string]interface{}, value interface{}, fields ...string) {
+// SetNestedField sets the value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedField(obj map[string]interface{}, value interface{}, fields ...string) bool {
```
I wished we had a debugging/strict-test mode where `value` is type checked. We use(d) to write arbitrary values into objects (https://github.com/kubernetes/kubernetes/pull/54803) and the compiler cannot help us with type checks.

ash2k replied:
I can send a follow up PR with a environment variable flag a-la https://github.com/kubernetes/kubernetes/blob/51e653dc68e47584ab571fc214a7aa5950918831/staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go#L81 and https://github.com/kubernetes/kubernetes/blob/ed423054ba0089a6d58fb0e048fe1acfd966f0d7/staging/src/k8s.io/client-go/tools/cache/mutation_detector.go#L34

sttts replied:
Would be great to run unit and maybe even integration tests with this.

ash2k replied:
Follow up in https://github.com/kubernetes/kubernetes/pull/55297

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -31,171 +31,278 @@ import (
+	delete(m, fields[len(fields)-1])
+}
+
+func getNestedString(obj map[string]interface{}, fields ...string) string {
```
nit: nestedStringOrDefault with `default string` arg?

ash2k replied:
`getNestedString()` is used in lots of places and default is an empty string everywhere. That is why I kept it as it is. Also I didn't want to touch the name because it would've increased the PR size even further. I can send a follow-up PR with rename if needed.

sttts replied:
Makes sense.

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -31,171 +31,278 @@ import (
-func setNestedSlice(obj map[string]interface{}, value []string, fields ...string) {
+// SetNestedStringSlice sets the string slice value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedStringSlice(obj map[string]interface{}, value []string, fields ...string) bool {
```
For the setters the bool return value feels unnatural.

nikhita replied:
Does an explicit error here look better? Something like that in https://github.com/kubernetes/kubernetes/pull/55168#discussion_r152580964?

sttts commented:
Having triaged issues in the old code recently, this PR is big step forward in cleaning up this area. Left a few comments, but I think we are not far from getting this in.

@liggitt @wojtek-t are you fine with the benchmarks? Any further concerns?

ash2k committed:
Useful helper functions for Unstructured

ash2k commented:
Rebased+squashed. I left first commit unchanged - it is just moving code around, no changes. Second one with all the changes.

sttts commented:
Waiting for @liggitt @wojtek-t's ok performance-wise. Will do a final review pass then.

ash2k commented:
/retest

deads2k commented:
> @liggitt @wojtek-t are you fine with the benchmarks? Any further concerns?

.08% change?  I don't think they can really be against it.

sttts commented:
@deads2k feel free to go through the code. More eyes help in this highly untyped and fragile area.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:
```diff
@@ -0,0 +1,470 @@
+	}
+
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	if len(gvk.Kind) == 0 {
```
Straight move?  I'd expect unstructured objects to always have a `gvk.Version` too.

ash2k replied:
Yes, I haven't touched this code. See https://github.com/kubernetes/kubernetes/pull/51940/commits/7aeaedd721604714adae4b9fb947dacd53821088#diff-31522f2d01b79a28cf7d5adaa03d5b26L685

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -478,31 +272,47 @@ func (u *Unstructured) GetDeletionTimestamp() *metav1.Time {
 func (u *Unstructured) SetDeletionTimestamp(timestamp *metav1.Time) {
 	if timestamp == nil {
-		u.setNestedField(nil, "metadata", "deletionTimestamp")
+		RemoveNestedField(u.Object, "metadata", "deletionTimestamp")
```
I'd expect a similar remove on `SetCreationTimestamp`

sttts replied:
Creation timestamp is not optional, at least the argument of the Go func is not. So at worst, we miss a feature here on the Go level.

deads2k commented:
> @deads2k feel free to go through the code. More eyes help in this highly untyped and fragile area.

It lgtm, but this area of the code is full of weird gotchas.  @sttts if you like it in the morning, I think it's fine.

/approve

sttts commented:
/lgtm

smarterclayton commented:
These shouldn't be in this package - I'm moving them into runtime, because meta/v1/unstructured is only for "typed" unstructured (i.e., anything that obeys our metadata rules), not generic to unstructured.

ash2k commented:
That makes sense. Thanks for an update.
